src/SourceGenerators/Uno.UI.SourceGenerators/XamlGenerator/XamlCodeGeneration.cs

```diff
@@ -288,6 +288,23 @@ private string GetSourceLink(Uno.Roslyn.MSBuildItem projectItemInstance)
 				var resourceKeys = GetResourceKeys(_generatorContext.CancellationToken);
 				var filesFull = new XamlFileParser(_excludeXamlNamespaces, _includeXamlNamespaces, _metadataHelper)
 					.ParseFiles(_xamlSourceFiles, _generatorContext.CancellationToken);
+
+				var xamlTypeToXamlTypeBaseMap = new Dictionary<INamedTypeSymbol, XamlRedirection.XamlType>();
+				foreach (var file in filesFull)
+				{
+					var topLevelControl = file.Objects.FirstOrDefault();
+					if (topLevelControl is null)
+					{
+						continue;
+					}
+
+					var xClassSymbol = XamlFileGenerator.FindClassSymbol(topLevelControl, _metadataHelper);
+					if (xClassSymbol is not null)
+					{
+						xamlTypeToXamlTypeBaseMap.Add(xClassSymbol, topLevelControl.Type);
+					}
+				}
```

Comment:
We need to make sure this doesn't hurt the performance. The reported issue is quite a niche case where we already produce a warning to guide the user towards an easy workaround (specify base type in codebehind).

If this hurts performance badly, I personally don't think we should take this change.

Reply:
Maybe we should be doing this in parallel, as there are many files to process.

Reply:
@jeromelaban Thanks! This has made a significant difference.